corehq/motech/dhis2/finders.py

```diff
@@ -0,0 +1,99 @@
+from collections import namedtuple
+from functools import partial
+
+from memoized import memoized_property
+
+from corehq.motech.dhis2.const import DHIS2_API_VERSION
+from corehq.motech.finders import MATCH_FUNCTIONS, MATCH_TYPE_EXACT
+
+CandidateScore = namedtuple('CandidateScore', 'candidate score')
+
+
+class TrackedEntityInstanceFinder:
+
+    def __init__(self, requests, case_config):
+        self.requests = requests
+        self.case_config = case_config
+
+    @property
+    def property_weights(self):
+        return self.case_config.finder_config.property_weights
+
+    @property
+    def confidence_margin(self):
+        return self.case_config.finder_config.confidence_margin
+
+    @memoized_property
+    def attr_type_id_value_source_by_case_property(self):
+        return {
+            value_source["case_property"]: (attr_type_id, value_source)
+            for attr_type_id, value_source in self.case_config.attributes.items()
+        }
+
+    def find_tracked_entity_instances(self, case_trigger_info):
+        """
+        Search DHIS2 for potential matches of the CommCare case. Score
+        search results and keep those with score > 1. If more than one
+        result has a score > 1, select the best candidate if it exceeds
+        a confidence margin. Otherwise return all results with score > 1.
+        """
+        results = self.fetch_query_results(case_trigger_info)
+        candidate_scores = []
+        for instance in results:
+            score = self.get_score(instance, case_trigger_info)
+            if score >= 1:
+                candidate_scores.append(CandidateScore(instance, score))
+
+        if len(candidate_scores) > 1:
+            candidate_scores = sorted(candidate_scores, key=lambda cs: cs.score, reverse=True)
+            if candidate_scores[0].score / candidate_scores[1].score > 1 + self.confidence_margin:
+                return [candidate_scores[0].candidate]
+        return [cs.candidate for cs in candidate_scores]
+
+    def fetch_query_results(self, case_trigger_info):
+        endpoint = f"/api/{DHIS2_API_VERSION}/trackedEntityInstances"
+        query_filters = self.get_query_filters(case_trigger_info)
+        if not query_filters:
+            return []
+        params = {
+            "ou": self.case_config.org_unit_id.get_value(case_trigger_info),
+            "filter": query_filters,
+            "ouMode": "DESCENDANTS",
+            "skipPaging": "true",
+        }
+        response = self.requests.get(endpoint, params=params, raise_for_status=True)
+        return response.json()["trackedEntityInstances"]
+
+    def get_query_filters(self, case_trigger_info):
+        filters = []
+        for property_weight in self.property_weights:
+            case_property = property_weight['case_property']
+            value = case_trigger_info.extra_fields[case_property]
+            if property_weight["match_type"] == MATCH_TYPE_EXACT and value:
```

Comment:
I assume `value` is expected to be an empty collection or `None`, which means "no value." Is it possible that it could be something like `0` (zero) or `''` (empty string) that evaluates to False and could/should be included in the query? In other words, would it make sense to do
```suggestion
            if property_weight["match_type"] == MATCH_TYPE_EXACT and value is not None:
```

Reply:
Nicely spotted, thank you. I believe that `""` is a value and `None` is not. But an API, its database, or its interface might not be that careful. How do you feel about considering None and empty strings as "not a value" and `0` as a value?

Reply:
Seems reasonable.